op.c

```diff
@@ -16685,26 +16751,115 @@ Perl_subsignature_finish(pTHX)
     yy_parser_signature *signature = PL_parser->signature;
     assert(signature);
 
-    OP *sigops = signature->elemops;
-    signature->elemops = NULL;
+    OP *sigops = NULL;
+
+    /* a nextstate right at the beginning */
+    sigops = op_append_elem(OP_LINESEQ, sigops,
+            newSTATEOP(0, NULL, NULL));
+
+    UV end_argix = signature->next_argix;
+
+    struct op_multiparam_aux *aux = (struct op_multiparam_aux *)PerlMemShared_malloc(
+            sizeof(struct op_multiparam_aux) + end_argix * sizeof(PADOFFSET));
+    aux->param_padix = (PADOFFSET *)((char *)aux + sizeof(struct op_multiparam_aux));
 
-    struct op_argcheck_aux *aux = (struct op_argcheck_aux *)
-        PerlMemShared_malloc( sizeof(struct op_argcheck_aux));
+    aux->min_args     = end_argix - signature->opt_params;
+    aux->n_positional = end_argix;
+    aux->slurpy       = signature->slurpy;
+    aux->slurpy_padix = 0;
 
-    aux->params     = signature->elems;
-    aux->opt_params = signature->optelems;
-    aux->slurpy     = signature->slurpy;
+    OP *multiparam = newUNOP_AUX(OP_MULTIPARAM, 0, NULL, (UNOP_AUX_item *)aux);
 
-    OP *check = newUNOP_AUX(OP_ARGCHECK, 0, NULL, (UNOP_AUX_item *)aux);
+    sigops = op_append_elem(OP_LINESEQ, sigops,
+            multiparam);
 
-    sigops = op_prepend_elem(OP_LINESEQ,
-            check,
-            sigops);
+    if(signature->fenceops) {
+        LISTOP *fenceops = cLISTOPx(signature->fenceops);
 
-    /* a nextstate right at the beginning */
-    sigops = op_prepend_elem(OP_LINESEQ,
-            newSTATEOP(0, NULL, NULL),
-            sigops);
+        /* Move the entire chain of kid ops in one go */
+        OpMORESIB_set(cLISTOPx(sigops)->op_last, fenceops->op_first);
+        cLISTOPx(sigops)->op_last = fenceops->op_last;
+        OpLASTSIB_set(cLISTOPx(sigops)->op_last, sigops);
```

Comment:
factor out `cLISTOPx(sigops)` to a C auto  i suspect its many derefs deep and keeps being re-read after each of these assignments in -O1/-O2. `cLISTOPx(sigops)->op_last` is written so many times its a puzzle what happened to the old target ptr struct, and the new target ptr, and is new or old target inside the container after this block.